### DISCUSSION TITLE
When will windows.system.userprofile.userprofilepersonalizationsettings UWP APIs be migrated to WinAppSDK

### DISCUSSION THREAD

**Discussion by divaypandey** [microsoft/WindowsAppSDK#3854] [General]
The [User Personalisation Settings](https://learn.microsoft.com/en-in/uwp/api/windows.system.userprofile.userprofilepersonalizationsettings?view=winrt-22621 ) are critical for applications to be able to change user Wallpaper/Lockscreen easily and were well supported in UWP

WinAppSDK which is supposed to be its successor has no support for it, and barely any communication for future plans to support it

Will it not be implemented at all?

**Comment by DarranRowe** [comment]
There is no documentation that this runtime class is not working in desktop applications. UserProfilePersonalizationSettings activates in a Windows API desktop application and IsSupported returns true. Are you having any particular issues when using it?

It can be confusing at times, but functionality in the UWP namespaces isn't limited to just the UWP application type. While there is still some inaccuracies [in the documentation](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/desktop-to-uwp-supported-api), Microsoft officially supports this. The WinUI 3 gallery even gives examples of using the file pickers from the UWP namespaces too.

**Reply by divaypandey** [reply]
So I tested further, in WinAppSDK 1.4 the `UserProfilePersonalizationSettings.IsSupported()` returns True

But, the `UserProfilePersonalizationSettings.Current.TrySetWallpaperImageAsync(file)` works only for _in-app local Storage_ files, i.e. the **"ms-appx:///local/img.jpg"** files

`StorageFile.GetFileFromApplicationUriAsync(localUri)` file when set as wallpaper/lockscreen, changes and works as intended
But, if the same StorageFile is procured from Path (_say a C drive's Downloads folder_):

`StorageFile.GetFileFromPathAsync(@"c:\some_path\img.jpg")`, its variable has access to the file correctly, but setting it as Wallpaper/LockScreen silently fails, ie `TrySetWallpaperImageAsync(file)` returns "False", there's no exception thrown

The same behaviour happens for `StorageFile.CreateStreamedFileFromUriAsync`, where too a temporary/ReadOnly file is visible in the variable when debugging, but setting it as wallpaper/lockscreen silently fails

For both failing scenarios I tried to turn them into Task objects and wait on it to see if its Exception is filled, but no error props up, it silently fails

**Reply by DarranRowe** [reply]
None of the documentation indicates that this is isn't available from an unpackaged desktop application. IsSupported returns true, however, the ms-appx scheme isn't available in an unpackaged desktop application. This means that in an unpackaged desktop application, StorageFile.GetFileFromPathAsync is the only option available.

There is no documentation stating that the path cannot or even must not be an absolute or relative filesystem path. So this is a toss up of this being a documentation issue or a bug in TrySetWallpaperImageAsync. This also doesn't fail silently, there is an error message passed to the debugger through RoOriginateError:
"Exception thrown at 0x00007FFA4D7C4C3C (KernelBase.dll) in BasicTestWindow.exe: WinRT originate error - 0x80073D54 : 'Error on internal call to AppThemeBrokeredStatics::SetThemeImage (Wallpaper)'."

This means that this is a difficult problem to comment further on. If this is intended to work in all desktop applications, and since this is a UWP API, the Windows App SDK is totally irrelevant, then this is a bug in Windows itself. If this is never meant to work then again, this is a bug in Windows itself since IsSupported shouldn't be returning true. The documentation should also point out that the path passed in needs to be a package relative URI. In both cases, this needs input from a Microsoft dev.
If this runtime class is intended to work with desktop applications then there doesn't really need to be a Windows App SDK implementation. All that needs to happen is the current implementation gets fixed. For now though, this needs input.

**Comment by castorix** [comment]
On my OS, Windows 10 22H2, I can change the LockScreen image with 
_Windows.System.UserProfile.LockScreen.SetImageFileAsync_ and a random path
but only in **x64** and it is wrong, even if I change it in "Settings > Personalization > Lock screen"...
